### ISSUE TITLE
Feature Request: Improve support for json, toml, xml, yaml and csv

### ISSUE THREAD

**Issue by samj1912** [updatecli/updatecli#870]
### Is your feature request related to a problem?

Currently it is not possible to use toml or csv files as source/condition/targets.

The filter/search capabilities for yaml are also not as expressive and do not allow specifying filter expressions to target keys.

### Solution you'd like

My proposal is to incorporate dasel - https://daseldocs.tomwright.me/use-as-a-go-package

a popular configuration parsing library into updatecli as a plugin to improve its capabilities.

It supports quite a few config file types https://daseldocs.tomwright.me/usage/supported-file-types

additionally it also supports more powerful search/query/put selector expressions -

- https://daseldocs.tomwright.me/selectors/next-available-index
- https://daseldocs.tomwright.me/selectors/all-indexes
- https://daseldocs.tomwright.me/selectors/dynamic
- https://daseldocs.tomwright.me/selectors/search
- https://daseldocs.tomwright.me/selectors/search-1

### Alternatives you've considered

Use libraries like yq but although they are powerful for yaml, dasel presents a better alternative since it supports more file types.

### Anything else?

Thank you for this amazing tool!

**Comment by samj1912** [comment]
Happy to work on this!

**Comment by olblak** [comment]
Hi @samj1912 ,

Thank you for the detailed issue.
First of all, I would definitely be interested to support standard file such as json,toml,csv,etc.
Having them would allow to easily implement "autodiscovery" for scenarios such as automatically bumping javascript dependencies specified in packages.json, or terraform,etc. 

Toml has been recurring demand, and I also need it in my projects. But due to lack of time, I didn't go further. So we went with the "file" resource + regex. It's not ideal and not easy to maintain but it works.
You can find more context on https://github.com/updatecli/updatecli/issues/534

Regarding Dasel, I was really enthusiast with when I discovered that project. I thought I would use it and contribute some of the the logic from Updatecli to it. hence I implemented the XML resource here 
-> https://github.com/updatecli/updatecli/pull/336/files

But then I realized they are many edge cases that I needed to fix first.
The XML part drops content.
The YAML drop comments and still use goyaml v2 so I would expect other unsupported edge cases. At this time, YAML support would reintroduce issues that I fixed in Updatecli but the query mechanism is more advanced so I am still keeping an eye on it.
The Toml and Json seems to be working fine but I didn't have the time to test further. In long term I wish I had more time to contribute to Dasel as it definitely makes sense to move some logic over there.

In short term, I focus on per resource. Several weeks ago, I discovered a great library to manipulate xml.

All the code for a plugin is located in `pkg/plugins/resource` for example
-> https://github.com/updatecli/updatecli/tree/main/pkg/plugins/resources/xml 

I recently added this xml plugin via this https://github.com/updatecli/updatecli/pull/823/files

Once the plugin implement the "resource" interface, then it has to be registered on https://github.com/updatecli/updatecli/blob/main/pkg/core/pipeline/resource/main.go#L47L125
So Updatecli knows how to map "kind" with a specific

**Comment by olblak** [comment]
I could have a look to this. 
I already have a branch with the XML and Dasel. I could easily adapt the code to experiment with json, or csv

**Comment by olblak** [comment]
@samj1912 I had already most of the code hanging in a stalled branch. I can also easily add csv support. I am not planning to integrate XML or YAML as several fixes are fix required

**Comment by olblak** [comment]
The problem that I have with Dasel is that it doesn't preserve "metadata" such as comments. So it's not an issue for json and csv, but it is for toml.

I had exactly the same issue when I looked at yaml and xml

**Comment by samj1912** [comment]
That's unfortunate. For toml it looks like the issue is with the underlying toml library that dasel uses.

**Comment by olblak** [comment]
Thanks @samj1912 for the suggestion, I had this Dasel integration in the back of my head for a quite long time and it ended being easier than I thought.

I added support for `json`/`toml`/`csv` in release [v0.34.0](https://github.com/updatecli/updatecli/releases/tag/v0.34.0)
I was really impressed by the power of advanced queries, so I'll try to find some time to improvement the YAML support as it could really benefit to Updatecli.

Regarding the multiple queries, I decided to add a parameter `multiple` to enable/disable multiple queries for  condition and target but I need additional testing to really understand if it's relevant or if I should just skip it.

The two drawback that I identified

1. Toml drop comments but we still have the possibility to use the resource file like I do [here](https://github.com/updatecli/website/blob/master/updatecli/updatecli.d/hugo.yaml#L51L59)
2. csv drop comments,

Meanwhile I consider this issue as solve
